ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Ledger.hs

```diff
@@ -0,0 +1,118 @@
+import           Ouroboros.Consensus.Protocol.Praos (Praos)
+import           Ouroboros.Consensus.Protocol.TPraos (TPraos)
+import           Ouroboros.Consensus.Shelley.Ledger (ShelleyBlock)
+import           Ouroboros.Consensus.TypeFamilyWrappers (WrapTxOut (..))
+
+instance CardanoHardForkConstraints c
+      => HasCanonicalTxIn (CardanoEras c) where
+  newtype instance CanonicalTxIn (CardanoEras c) = CardanoTxIn {
+      getCardanoTxIn :: SL.TxIn c
+    }
+    deriving stock (Show, Eq, Ord)
+    deriving newtype NoThunks
+
+  injectCanonicalTxIn IZ       byronTxIn   = absurd byronTxIn
+  injectCanonicalTxIn (IS idx) shelleyTxIn = case idx of
+      IZ                               -> CardanoTxIn shelleyTxIn
+      IS IZ                            -> CardanoTxIn shelleyTxIn
+      IS (IS IZ)                       -> CardanoTxIn shelleyTxIn
+      IS (IS (IS IZ))                  -> CardanoTxIn shelleyTxIn
+      IS (IS (IS (IS IZ)))             -> CardanoTxIn shelleyTxIn
+      IS (IS (IS (IS (IS IZ))))        -> CardanoTxIn shelleyTxIn
+      IS (IS (IS (IS (IS (IS idx'))))) -> case idx' of {}
+
+  distribCanonicalTxIn IZ _                 =
+      error "distribCanonicalTxIn: Byron has no TxIns"
+  distribCanonicalTxIn (IS idx) cardanoTxIn = case idx of
+      IZ                               -> getCardanoTxIn cardanoTxIn
+      IS IZ                            -> getCardanoTxIn cardanoTxIn
+      IS (IS IZ)                       -> getCardanoTxIn cardanoTxIn
+      IS (IS (IS IZ))                  -> getCardanoTxIn cardanoTxIn
+      IS (IS (IS (IS IZ)))             -> getCardanoTxIn cardanoTxIn
+      IS (IS (IS (IS (IS IZ))))        -> getCardanoTxIn cardanoTxIn
+      IS (IS (IS (IS (IS (IS idx'))))) -> case idx' of {}
+
+  encodeCanonicalTxIn   = Core.toEraCBOR @(ShelleyEra c) . getCardanoTxIn
+
+  decodeCanonicalTxIn = CardanoTxIn <$> Core.fromEraCBOR @(ShelleyEra c)
+
+data CardanoTxOut c =
+    ShelleyTxOut !(Value (LedgerState (ShelleyBlock (TPraos c) (ShelleyEra c))))
+  | AllegraTxOut !(Value (LedgerState (ShelleyBlock (TPraos c) (AllegraEra c))))
+  | MaryTxOut    !(Value (LedgerState (ShelleyBlock (TPraos c) (MaryEra c))))
+  | AlonzoTxOut  !(Value (LedgerState (ShelleyBlock (TPraos c) (AlonzoEra c))))
+  | BabbageTxOut !(Value (LedgerState (ShelleyBlock (Praos c) (BabbageEra c))))
+  | ConwayTxOut  !(Value (LedgerState (ShelleyBlock (Praos c) (ConwayEra c))))
+  deriving stock (Show, Eq, Generic)
+  deriving anyclass NoThunks
+
+toNS :: CardanoTxOut c -> NS WrapTxOut (CardanoEras c)
+toNS = \case
+    ShelleyTxOut txout ->                     S $ Z $ WrapTxOut txout
+    AllegraTxOut txout ->                 S $ S $ Z $ WrapTxOut txout
+    MaryTxOut txout    ->             S $ S $ S $ Z $ WrapTxOut txout
+    AlonzoTxOut txout  ->         S $ S $ S $ S $ Z $ WrapTxOut txout
+    BabbageTxOut txout ->     S $ S $ S $ S $ S $ Z $ WrapTxOut txout
+    ConwayTxOut txout  -> S $ S $ S $ S $ S $ S $ Z $ WrapTxOut txout
```

Comment:
Would there be any benefit in using `S (S (...` instead of `S $ S $ ...`?

Reply:
I think it comes down to the same code eventually